File: FinderAndroid/app/src/main/java/com/finder/app/services/WebSocketService.kt
package com.finder.app.services

import com.finder.app.models.ServerMessage
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.asSharedFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import kotlinx.serialization.json.Json
import okhttp3.*
import org.json.JSONObject
import java.util.concurrent.TimeUnit

object WebSocketService {
    private const val WS_URL = "ws://155.212.165.134:3001"

    private var webSocket: WebSocket? = null
    private val client = OkHttpClient.Builder()
        .readTimeout(0, TimeUnit.MILLISECONDS)
        .pingInterval(30, TimeUnit.SECONDS)
        .build()
    private val json = Json { ignoreUnknownKeys = true; coerceInputValues = true }
    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.Main)

    private val _isConnected = MutableStateFlow(false)
    val isConnected = _isConnected.asStateFlow()

    // Events
    private val _messageReceived = MutableSharedFlow<ServerMessage>(extraBufferCapacity = 64)
    val messageReceived = _messageReceived.asSharedFlow()

    private val _typingReceived = MutableSharedFlow<Pair<String, String>>(extraBufferCapacity = 64) // chatId, userId
    val typingReceived = _typingReceived.asSharedFlow()

    private val _userStatusChanged = MutableSharedFlow<Pair<String, Boolean>>(extraBufferCapacity = 64) // userId, isOnline
    val userStatusChanged = _userStatusChanged.asSharedFlow()

    private val _incomingCall = MutableSharedFlow<IncomingCallData>(extraBufferCapacity = 16)
    val incomingCall = _incomingCall.asSharedFlow()

    private val _callEnded = MutableSharedFlow<String>(extraBufferCapacity = 16)
    val callEnded = _callEnded.asSharedFlow()

    // WebRTC signaling
    private val _webrtcOffer = MutableSharedFlow<WebRTCSignal>(extraBufferCapacity = 16)
    val webrtcOffer = _webrtcOffer.asSharedFlow()

    private val _webrtcAnswer = MutableSharedFlow<WebRTCSignal>(extraBufferCapacity = 16)
    val webrtcAnswer = _webrtcAnswer.asSharedFlow()

    private val _iceCandidate = MutableSharedFlow<WebRTCSignal>(extraBufferCapacity = 16)
    val iceCandidate = _iceCandidate.asSharedFlow()

    private var token: String? = null
    private var reconnectAttempts = 0

    fun connect(authToken: String) {
        token = authToken
        reconnectAttempts = 0
        doConnect()
    }

    private fun doConnect() {
        val request = Request.Builder().url(WS_URL).build()
        webSocket = client.newWebSocket(request, object : WebSocketListener() {
            override fun onOpen(webSocket: WebSocket, response: Response) {
                _isConnected.value = true
                reconnectAttempts = 0
                // Authenticate
                val authMsg = JSONObject().apply {
                    put("type", "auth")
                    put("token", token)
                }
                webSocket.send(authMsg.toString())
            }

            override fun onMessage(webSocket: WebSocket, text: String) {
                handleMessage(text)
            }

            override fun onClosing(webSocket: WebSocket, code: Int, reason: String) {
                webSocket.close(1000, null)
                _isConnected.value = false
            }

            override fun onFailure(webSocket: WebSocket, t: Throwable, response: Response?) {
                _isConnected.value = false
                attemptReconnect()
            }
        })
    }

    fun disconnect() {
        webSocket?.close(1000, "User disconnect")
        webSocket = null
        _isConnected.value = false
    }

    // Send methods
    fun sendMessage(chatId: String, text: String, messageType: String = "text") {
        val msg = JSONObject().apply {
            put("type", "message")
            put("chat_id", chatId)
            put("text", text)
            put("message_type", messageType)
        }
        webSocket?.send(msg.toString())
    }

    fun sendTyping(chatId: String) {
        val msg = JSONObject().apply {
            put("type", "typing")
            put("chat_id", chatId)
        }
        webSocket?.send(msg.toString())
    }

    fun sendRead(chatId: String, messageId: String) {
        val msg = JSONObject().apply {
            put("type", "read")
            put("chat_id", chatId)
            put("message_id", messageId)
        }
        webSocket?.send(msg.toString())
    }

    fun sendWebRTCOffer(callId: String, sdp: String, to: String) {
        val msg = JSONObject().apply {
            put("type", "webrtc-offer")
            put("call_id", callId)
            put("sdp", sdp)
            put("to", to)
        }
        webSocket?.send(msg.toString())
    }

    fun sendWebRTCAnswer(callId: String, sdp: String, to: String) {
        val msg = JSONObject().apply {
            put("type", "webrtc-answer")
            put("call_id", callId)
            put("sdp", sdp)
            put("to", to)
        }
        webSocket?.send(msg.toString())
    }

    fun sendICECandidate(callId: String, candidate: String, to: String) {
        val msg = JSONObject().apply {
            put("type", "ice-candidate")
            put("call_id", callId)
            put("candidate", candidate)
            put("to", to)
        }
        webSocket?.send(msg.toString())
    }

    private fun handleMessage(text: String) {
        try {
            val jsonObj = JSONObject(text)
            val type = jsonObj.optString("type")
            scope.launch {
                when (type) {
                    "message" -> {
                        val msgJson = jsonObj.optJSONObject("message")?.toString() ?: jsonObj.toString()
                        val serverMsg = json.decodeFromString<ServerMessage>(msgJson)
                        _messageReceived.emit(serverMsg)
                    }
                    "typing" -> {
                        val chatId = jsonObj.optString("chat_id")
                        val userId = jsonObj.optString("user_id")
                        if (chatId.isNotEmpty() && userId.isNotEmpty()) {
                            _typingReceived.emit(chatId to userId)
                        }
                    }
                    "read" -> { /* handled via API */ }
                    "user-online" -> {
                        val userId = jsonObj.optString("user_id")
                        if (userId.isNotEmpty()) _userStatusChanged.emit(userId to true)
                    }
                    "user-offline" -> {
                        val userId = jsonObj.optString("user_id")
                        if (userId.isNotEmpty()) _userStatusChanged.emit(userId to false)
                    }
                    "webrtc-offer" -> {
                        _webrtcOffer.emit(WebRTCSignal(
                            callId = jsonObj.optString("call_id"),
                            sdp = jsonObj.optString("sdp"),
                            from = jsonObj.optString("from")
                        ))
                    }
                    "webrtc-answer" -> {
                        _webrtcAnswer.emit(WebRTCSignal(
                            callId = jsonObj.optString("call_id"),
                            sdp = jsonObj.optString("sdp"),
                            from = jsonObj.optString("from")
                        ))
                    }
                    "ice-candidate" -> {
                        _iceCandidate.emit(WebRTCSignal(
                            callId = jsonObj.optString("call_id"),
                            sdp = jsonObj.optString("candidate"),
                            from = jsonObj.optString("from")
                        ))
                    }
                    "call-incoming" -> {
                        _incomingCall.emit(IncomingCallData(
                            callId = jsonObj.optString("call_id"),
                            chatId = jsonObj.optString("chat_id"),
                            isVideo = jsonObj.optBoolean("is_video"),
                            from = jsonObj.optString("from")
                        ))
                    }
                    "call-ended" -> {
                        val callId = jsonObj.optString("call_id")
                        if (callId.isNotEmpty()) _callEnded.emit(callId)
                    }
                }
            }
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    private fun attemptReconnect() {
        if (reconnectAttempts >= 10) return
        reconnectAttempts++
        val delay = minOf(reconnectAttempts * 2000L, 30000L)
        scope.launch {
            kotlinx.coroutines.delay(delay)
            if (!_isConnected.value && token != null) {
                doConnect()
            }
        }
    }
}

data class IncomingCallData(
    val callId: String,
    val chatId: String,
    val isVideo: Boolean,
    val from: String
)

data class WebRTCSignal(
    val callId: String,
    val sdp: String,
    val from: String
)

File: Finder/Services/WebSocketService.swift
import Foundation
import Combine

// MARK: - WebSocket Message Types

enum WSIncoming {
    case message(ServerMessage)
    case typing(chatId: String, userId: String)
    case read(chatId: String, userId: String, messageId: String)
    case userOnline(userId: String)
    case userOffline(userId: String)
    case webrtcOffer(callId: String, sdp: String, from: String)
    case webrtcAnswer(callId: String, sdp: String, from: String)
    case webrtcIceCandidate(callId: String, candidate: String, from: String)
    case callIncoming(callId: String, chatId: String, isVideo: Bool, from: String)
    case callEnded(callId: String)
    case unknown
}

class WebSocketService: ObservableObject {
    static let shared = WebSocketService()

    private let wsURL = "ws://155.212.165.134:3001"
    private var webSocket: URLSessionWebSocketTask?
    private var session: URLSession?
    private var pingTimer: Timer?

    @Published var isConnected = false

    // Event publishers
    let messageReceived = PassthroughSubject<ServerMessage, Never>()
    let typingReceived = PassthroughSubject<(chatId: String, userId: String), Never>()
    let readReceived = PassthroughSubject<(chatId: String, userId: String, messageId: String), Never>()
    let userStatusChanged = PassthroughSubject<(userId: String, isOnline: Bool), Never>()
    let webrtcSignal = PassthroughSubject<WSIncoming, Never>()
    let incomingCall = PassthroughSubject<(callId: String, chatId: String, isVideo: Bool, from: String), Never>()
    let callEnded = PassthroughSubject<String, Never>()

    private var reconnectAttempts = 0
    private let maxReconnectAttempts = 10

    private init() {}

    // MARK: - Connect

    func connect(token: String) {
        guard let url = URL(string: wsURL) else { return }

        let config = URLSessionConfiguration.default
        session = URLSession(configuration: config)
        webSocket = session?.webSocketTask(with: url)
        webSocket?.resume()

        // Authenticate
        let authMsg: [String: String] = ["type": "auth", "token": token]
        send(authMsg)

        startReceiving()
        startPing()

        reconnectAttempts = 0
        DispatchQueue.main.async { self.isConnected = true }
    }

    func disconnect() {
        pingTimer?.invalidate()
        pingTimer = nil
        webSocket?.cancel(with: .normalClosure, reason: nil)
        webSocket = nil
        DispatchQueue.main.async { self.isConnected = false }
    }

    // MARK: - Send

    func sendMessage(chatId: String, text: String, messageType: String = "text") {
        let msg: [String: String] = [
            "type": "message",
            "chat_id": chatId,
            "text": text,
            "message_type": messageType
        ]
        send(msg)
    }

    func sendTyping(chatId: String) {
        send(["type": "typing", "chat_id": chatId])
    }

    func sendRead(chatId: String, messageId: String) {
        send(["type": "read", "chat_id": chatId, "message_id": messageId])
    }

    // MARK: - WebRTC Signaling

    func sendWebRTCOffer(callId: String, sdp: String, to: String) {
        send([
            "type": "webrtc-offer",
            "call_id": callId,
            "sdp": sdp,
            "to": to
        ])
    }

    func sendWebRTCAnswer(callId: String, sdp: String, to: String) {
        send([
            "type": "webrtc-answer",
            "call_id": callId,
            "sdp": sdp,
            "to": to
        ])
    }

    func sendICECandidate(callId: String, candidate: String, to: String) {
        send([
            "type": "ice-candidate",
            "call_id": callId,
            "candidate": candidate,
            "to": to
        ])
    }

    // MARK: - Internal

    private func send(_ dict: [String: String]) {
        guard let data = try? JSONSerialization.data(withJSONObject: dict),
              let str = String(data: data, encoding: .utf8) else { return }

        webSocket?.send(.string(str)) { error in
            if let error = error {
                print("[WS] Send error: \(error.localizedDescription)")
            }
        }
    }

    private func startReceiving() {
        webSocket?.receive { [weak self] result in
            switch result {
            case .success(let message):
                switch message {
                case .string(let text):
                    self?.handleMessage(text)
                case .data(let data):
                    if let text = String(data: data, encoding: .utf8) {
                        self?.handleMessage(text)
                    }
                @unknown default:
                    break
                }
                // Continue receiving
                self?.startReceiving()

            case .failure(let error):
                print("[WS] Receive error: \(error.localizedDescription)")
                DispatchQueue.main.async { self?.isConnected = false }
                self?.attemptReconnect()
            }
        }
    }

    private func handleMessage(_ text: String) {
        guard let data = text.data(using: .utf8),
              let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let type = json["type"] as? String else { return }

        DispatchQueue.main.async { [self] in
            switch type {
            case "message":
                if let msgData = try? JSONSerialization.data(withJSONObject: json["message"] ?? json),
                   let serverMsg = try? JSONDecoder().decode(ServerMessage.self, from: msgData) {
                    messageReceived.send(serverMsg)
                }

            case "typing":
                if let chatId = json["chat_id"] as? String,
                   let userId = json["user_id"] as? String {
                    typingReceived.send((chatId: chatId, userId: userId))
                }

            case "read":
                if let chatId = json["chat_id"] as? String,
                   let userId = json["user_id"] as? String,
                   let messageId = json["message_id"] as? String {
                    readReceived.send((chatId: chatId, userId: userId, messageId: messageId))
                }

            case "user-online":
                if let userId = json["user_id"] as? String {
                    userStatusChanged.send((userId: userId, isOnline: true))
                }

            case "user-offline":
                if let userId = json["user_id"] as? String {
                    userStatusChanged.send((userId: userId, isOnline: false))
                }

            case "webrtc-offer":
                if let callId = json["call_id"] as? String,
                   let sdp = json["sdp"] as? String,
                   let from = json["from"] as? String {
                    webrtcSignal.send(.webrtcOffer(callId: callId, sdp: sdp, from: from))
                }

            case "webrtc-answer":
                if let callId = json["call_id"] as? String,
                   let sdp = json["sdp"] as? String,
                   let from = json["from"] as? String {
                    webrtcSignal.send(.webrtcAnswer(callId: callId, sdp: sdp, from: from))
                }

            case "ice-candidate":
                if let callId = json["call_id"] as? String,
                   let candidate = json["candidate"] as? String,
                   let from = json["from"] as? String {
                    webrtcSignal.send(.webrtcIceCandidate(callId: callId, candidate: candidate, from: from))
                }

            case "call-incoming":
                if let callId = json["call_id"] as? String,
                   let chatId = json["chat_id"] as? String,
                   let isVideo = json["is_video"] as? Bool,
                   let from = json["from"] as? String {
                    incomingCall.send((callId: callId, chatId: chatId, isVideo: isVideo, from: from))
                }

            case "call-ended":
                if let callId = json["call_id"] as? String {
                    callEnded.send(callId)
                }

            default:
                break
            }
        }
    }

    private func startPing() {
        pingTimer?.invalidate()
        pingTimer = Timer.scheduledTimer(withTimeInterval: 30, repeats: true) { [weak self] _ in
            self?.webSocket?.sendPing { error in
                if let error = error {
                    print("[WS] Ping error: \(error.localizedDescription)")
                    DispatchQueue.main.async { self?.isConnected = false }
                    self?.attemptReconnect()
                }
            }
        }
    }

    private func attemptReconnect() {
        guard reconnectAttempts < maxReconnectAttempts else {
            print("[WS] Max reconnect attempts reached")
            return
        }

        reconnectAttempts += 1
        let delay = min(Double(reconnectAttempts) * 2, 30)

        DispatchQueue.main.asyncAfter(deadline: .now() + delay) { [weak self] in
            guard let token = NetworkService.shared.authToken else { return }
            print("[WS] Reconnecting... attempt \(self?.reconnectAttempts ?? 0)")
            self?.connect(token: token)
        }
    }
}
